src/dmd/lexer.d

```diff
@@ -1805,6 +1826,7 @@ class Lexer
         {
             error("unterminated character constant");
             t.unsvalue = '?';
+            errorRecovery();
```

Comment:
Thanks, looks better now.

As you also noted, the other error case above doesn't need extra recovery, so the local function is only called once. It's probably simpler to just add the code here. Sorry for the misleading hint.

There's still some mishandling of line breaks, i.e. unicode characters LS and PS. If the first character is a line break an error is emitted above, so I wouldn't try to find other characters beyond the end of the line. Here's a possible implementation:
```D
        if (*p != '\'')
        {
            // try to recover: stop at '\'', end of file, line break
            //  or characters probably important to recover parsing
            while (*p != '\'' && *p != 0x1A && *p != 0 && *p != '\n' && *p != '\r'
                   && *p != ';' && *p != ')' && *p != ']' && *p != '}')
            {
                if (*p & 0x80)
                {
                    const s = p;
                    c = decodeUTF();
                    if (c == LS || c == PS)
                    {
                        p = s;
                        break;
                    }
                }
                p++;
            }

            if (*p == '\'')
            {
                error("character constant has multiple characters");
                p++;
            }
            else
                error("unterminated character constant");
            t.unsvalue = '?';
            return tk;
        }
```

Reply:
It's important **not** to stop at the end of the line. The code above fails on this test: https://github.com/dlang/dmd/blob/57cd63e3095712b02b2831ced1261a053b84bd41/test/fail_compilation/lexer4.d#L22. 

However, I found your suggestion very useful so I adapted it to handle the case above. Thanks!

Reply:
> It's important **not** to stop at the end of the line. The code above fails on this test:
> 
> https://github.com/dlang/dmd/blob/57cd63e3095712b02b2831ced1261a053b84bd41/test/fail_compilation/lexer4.d#L22

This case is already handled here:
https://github.com/dlang/dmd/blob/607cae0fd355775d40786b5293808e453773db7c/src/dmd/lexer.d#L1773-L1776

and the \r handling is similar to what I propose (let the lexer handle the newline before the next token): https://github.com/dlang/dmd/blob/607cae0fd355775d40786b5293808e453773db7c/src/dmd/lexer.d#L1777-L1778

I still think stopping at newlines is better, otherwise code like this only recovers five lines later:
```D
dchar ch =';

while(ch & 0x80 == 0x80 &&
          condition ||
          ch & 0xc0 == 0xc0 &&
          other_condition)
{
}
```

Reply:
Yes, you are right. Thanks